rundeckapp/grails-app/services/rundeck/services/asyncimport/AsyncImportService.groovy

```diff
@@ -0,0 +1,1073 @@
+    Boolean statusFileExists(String projectName){
+        try {
+            def fwkProject = frameworkService.getFrameworkProject(projectName)
+            def statusFilepath = "${JSON_FILE_PREFIX}${projectName}${JSON_FILE_EXT}"
+            if( !fwkProject.existsFileResource(statusFilepath) ){
+                return false
+            }
+            return true
+        } catch (Exception e) {
+            e.printStackTrace();
+            throw e
+        }
+    }
+
+    /**
+     *
+     * Gets the status file content as an object.
+     *
+     * @param projectName - Required param to project.
+     *
+     * @return AsyncImportStatusDTO - DTO with all the status file content.
+     */
+    @Override
+    AsyncImportStatusDTO getAsyncImportStatusForProject(String projectName, ByteArrayOutputStream out = null) {
+        try{
+            if( out == null ){
+                out = new ByteArrayOutputStream()
+            }
+            final def fwkProject = frameworkService.getFrameworkProject(projectName)
+            fwkProject.loadFileResource(JSON_FILE_PREFIX + projectName + JSON_FILE_EXT, out)
+            def obj = new JsonSlurper().parseText(out.toString()) as AsyncImportStatusDTO
+            log.debug("Async Import status file content: ${obj.toString()}")
+            return obj
+        }catch(Exception e){
+            log.error("Error during the async import file extraction process: ${e.message}")
+            throw e
+        }
+    }
+
+    /**
+     * Saves a new or update a status file in db. To create a new status file, pass a project name, to update one,
+     * pass a new AsyncImportStatusDTO with null unchanged properties, leaving only the updated properties non-null.
+     *
+     * @param projectName (nullable)
+     * @param newStatus (nullable)
+     * @return Long bytes written.
+     */
+    Long saveAsyncImportStatusForProject(String projectName = null, AsyncImportStatusDTO newStatus = null){
+        def resource
+        def statusPersist
+
+        if( !projectName && !newStatus ){
+            throw new AsyncImportException("Must pass either a projectName or a new AsyncImportStatusDTO in method call.")
+        }
+
+        try {
+            if( newStatus != null ){ // update scenario
+                statusPersist = new AsyncImportStatusDTO(newStatus)
+            }else{
+                statusPersist = new AsyncImportStatusDTO(
+                        projectName,
+                        AsyncImportMilestone.M1_CREATED.milestoneNumber
+                ).with {
+                    it.lastUpdated = new Date().toString()
+                    it.lastUpdate = AsyncImportMilestone.M1_CREATED.name
+                    return it
+                }
+            }
+            def jsonStatus = statusPersist as JSON
+            def inputStream = new ByteArrayInputStream(jsonStatus.toString().getBytes(StandardCharsets.UTF_8));
+            if (!statusPersist.projectName || statusPersist.projectName.size() <= 0) {
```

Comment:
is it even possible to have an status without projectName?

Reply:
Yes, we can. This is a validation to assure the right behavior.